### PR TITLE
Initial X25519KeyPair implementation.

### PR DESCRIPTION
**Pull request by dmitrizagidulin**

### CODE REVIEW

**Comment by dmitrizagidulin** [comment]
Ready for review.

File: lib/X25519KeyPair.js [hunk 1]
```diff
@@ -0,0 +1,318 @@
...
+const forge = require('node-forge');
+const {util: {binary: {base58}}} = forge;
+const {LDKeyPair} = require('crypto-ld');
+const {base58Decode} = require('crypto-ld/lib/util');
+const ed2curve = require('ed2curve');
+const nacl = require('tweetnacl');
+
+class X25519KeyPair extends LDKeyPair {
+  /**
+   * An implementation of x25519-xsalsa20-poly1305
```
Comment (dlongley) [inline lib/X25519KeyPair.js]:
```suggestion
   * An implementation of x25519.
```

File: lib/X25519KeyPair.js [hunk 2]
```diff
@@ -0,0 +1,318 @@
...
+   * X25519KeyPair { ...
+   * @param {KeyPairOptions} [options={}]
+   * @param {string} [options.controller]
+   * @param {string} [options.id]
+   *
+   * @returns {Promise<X25519KeyPair>} Generates a key pair.
+   */
+  static async generate(options = {}) {
+    // Each is a Uint8Array with 32-byte key
+    const {publicKey, secretKey: privateKey} = nacl.box.keyPair()
```
Comment (dlongley) [inline lib/X25519KeyPair.js]:
```suggestion
    const {publicKey, secretKey: privateKey} = nacl.box.keyPair();
```

My guess is this isn't being linted...

File: lib/X25519KeyPair.js [hunk 3]
```diff
@@ -0,0 +1,318 @@
...
+
+    return new X25519KeyPair({
+      publicKeyBase58: base58.encode(publicKey),
+      privateKeyBase58: base58.encode(privateKey),
+      ...options
+    });
+  }
+
+  /**
+   * Creates an X25519KeyPair Key Pair from an existing key (constructor method)
```
Comment (dlongley) [inline lib/X25519KeyPair.js]:
```suggestion
   * Creates an X25519KeyPair Key Pair from an existing key (constructor method).
```

File: lib/X25519KeyPair.js [hunk 4]
```diff
@@ -0,0 +1,318 @@
...
+      xKey.privateKeyBase58 = X25519KeyPair
+        .convertFromEdPrivateKey(edKeyPair.privateKeyBase58)
+    }
+
+    return xKey;
+  }
+
+  /**
+   * @param {string} edPublicKeyBase58 - base58 encoded Ed25519 Public key
+   * @returns {string} base58 encoded X25519 Public key
```
Comment (dlongley) [inline lib/X25519KeyPair.js]: Linting stuff again; inconsistent use of `-`, no periods, etc.

File: lib/X25519KeyPair.js [hunk 5]
```diff
@@ -0,0 +1,318 @@
...
+      keyMaterial: edPublicKeyBase58,
+      type: 'public'
+    });
+
+    // Converts a 32-byte Ed25519 public key into a 32-byte Curve25519 key
+    // Returns null if the given public key in not a valid Ed25519 public key.
+    const dhPubkeyBytes = ed2curve.convertPublicKey(edPubkeyBytes);
+    if(!dhPubkeyBytes) {
+      throw new Error(
+        'Error converting to X25519: Invalid Ed25519 public key.');
```
Comment (dlongley) [inline lib/X25519KeyPair.js]:
```suggestion
        'Error converting to X25519; Invalid Ed25519 public key.');
```

File: lib/X25519KeyPair.js [hunk 6]
```diff
@@ -0,0 +1,318 @@
...
+      decode: base58.decode,
+      keyMaterial: edPrivateKeyBase58,
+      type: 'private'
+    });
+    // Converts a 64-byte Ed25519 secret key (or just the first 32-byte part of
+    // it, which is the secret value) into a 32-byte Curve25519 secret key
+    const dhPrivkeyBytes = ed2curve.convertSecretKey(edPrivkeyBytes);
+    if(!dhPrivkeyBytes) {
+      throw new Error(
+        'Error converting to X25519: Invalid Ed25519 private key.');
```
Comment (dlongley) [inline lib/X25519KeyPair.js]:
```suggestion
        'Error converting to X25519; Invalid Ed25519 private key.');
```

File: lib/X25519KeyPair.js (hunk 1 above)
Comment (dmitrizagidulin) [reply lib/X25519KeyPair.js]: Fixed, thanks.

File: lib/X25519KeyPair.js (hunk 2 above)
Comment (dmitrizagidulin) [reply lib/X25519KeyPair.js]: Fixed.

File: lib/X25519KeyPair.js (hunk 3 above)
Comment (dmitrizagidulin) [reply lib/X25519KeyPair.js]: Fixed.

File: lib/X25519KeyPair.js (hunk 4 above)
Comment (dmitrizagidulin) [reply lib/X25519KeyPair.js]: Fixed.

File: lib/X25519KeyPair.js (hunk 5 above)
Comment (dmitrizagidulin) [reply lib/X25519KeyPair.js]: Changed.

**Comment by dmitrizagidulin** [comment]
@dlongley I've addressed your changes; let me know if there's anything, and let's merge & publish this?

File: lib/X25519KeyPair.js [hunk 7]
```diff
@@ -0,0 +1,321 @@
...
+const {util: {binary: {base58}}} = forge;
+const {LDKeyPair} = require('crypto-ld');
+const {base58Decode} = require('crypto-ld/lib/util');
+const ed2curve = require('ed2curve');
+const nacl = require('tweetnacl');
+
+class X25519KeyPair extends LDKeyPair {
+  /**
+   * An implementation of x25519
+   * [X25519 Key Agreement 2019]{@link https://w3c-dvcg.github.io/} representation
```
Comment (davidlehn) [inline lib/X25519KeyPair.js]: Is that the best link?  Not clear what is x25519 related there.

File: lib/X25519KeyPair.js [hunk 8]
```diff
@@ -0,0 +1,321 @@
...
+    this.privateKeyBase58 = options.privateKeyBase58;
+    this.publicKeyBase58 = options.publicKeyBase58;
+  }
+  /**
+   * Returns the Base58 encoded public key.
+   * @implements {LDKeyPair#publicKey}
+   * @readonly
+   *
+   * @returns {string} The Base58 encoded public key.
+   * @see [publicKey]{@link ./LDKeyPair.md#publicKey}
```
Comment (davidlehn) [inline lib/X25519KeyPair.js]: Should these links, here and below, to crypto-ld docs be specified some other way?

Comment (dmitrizagidulin) [reply lib/X25519KeyPair.js]: I'm honestly not sure, I didn't put them in there. I think it's advanced JSDoc usage?

File: lib/X25519KeyPair.js (hunk 7 above)
Comment (dmitrizagidulin) [reply lib/X25519KeyPair.js]: I suspect that link was put in there in an aspirational capacity - as a reminder to create an X25519 sig suite repo, in that org?

File: lib/X25519KeyPair.js [hunk 9]
```diff
@@ -0,0 +1,321 @@
...
+        error: new Error('`fingerprint` must be a multibase encoded string.'),
+        valid: false
+      };
+    }
+    let fingerprintBuffer;
+    try {
+      fingerprintBuffer = base58Decode({
+        decode: base58.decode,
+        keyMaterial: fingerprint.slice(1),
+        type: `fingerprint's`
```
Comment (davidlehn) [inline lib/X25519KeyPair.js]: I'll just note that's an odd looking "type".

Comment (dmitrizagidulin) [reply lib/X25519KeyPair.js]: Agreed. I wasn't sure what was going on there.

Comment (davidlehn) [reply lib/X25519KeyPair.js]: Oh... I see that's just for debugging.

File: lib/X25519KeyPair.js [hunk 10]
```diff
@@ -0,0 +1,321 @@
...
+        decode: base58.decode,
+        keyMaterial: this.publicKeyBase58,
+        type: 'public'
+      });
+    } catch(e) {
+      return {error: e, valid: false};
+    }
+
+    // validate the first multicodec byte 0x25
+    const valid = fingerprintBuffer.slice(0, 1).toString('hex') === '25' &&
```
Comment (davidlehn) [inline lib/X25519KeyPair.js]: Why the slice/hex?  Just use [0] == 0x25.  That's done in fromFingerprint.

Comment (dmitrizagidulin) [reply lib/X25519KeyPair.js]: Good call.

File: lib/X25519KeyPair.js (hunk 8 above)
Comment (davidlehn) [reply lib/X25519KeyPair.js]: I assumed they were copied from crypto-ld code that does have that md file.

Comment (dmitrizagidulin) [reply lib/X25519KeyPair.js]: Ah, ok. Will remove.

File: lib/X25519KeyPair.js (hunk 10 above)
Comment (dmitrizagidulin) [reply lib/X25519KeyPair.js]: Changed.

**Pull request merged**